framework/src/meshgenerators/XYZDelaunayGenerator.C

```diff
@@ -0,0 +1,408 @@
+      "hole_boundaries",
+      "Boundary names to set on holes.  Default IDs are numbered up from 1 if no hole meshes are "
+      "stitched; or from maximum boundary ID of all the stitched hole meshes + 2.");
+
+  params.addParam<bool>("smooth_triangulation",
+                        false,
+                        "Whether to do Laplacian mesh smoothing on the generated triangles.");
+  params.addParam<std::vector<MeshGeneratorName>>(
+      "holes", std::vector<MeshGeneratorName>(), "The MeshGenerators that define mesh holes.");
+  params.addParam<std::vector<bool>>(
+      "stitch_holes", std::vector<bool>(), "Whether to stitch to the mesh defining each hole.");
+
+  params.addRangeCheckedParam<Real>(
+      "desired_volume",
+      0,
+      "desired_volume>=0",
+      "Desired (maximum) tetrahedral volume, or 0 to skip uniform refinement");
+
+  params.addParam<MooseEnum>(
+      "algorithm",
+      algorithm,
+      "Control the use of binary search for the nodes of the stitched surfaces.");
+  params.addParam<bool>(
+      "verbose_stitching", false, "Whether mesh stitching should have verbose output.");
+
+  params.addClassDescription(
+      "Creates tetrahedral 3D meshes within boundaries defined by input meshes.");
+
+  return params;
+}
+
+XYZDelaunayGenerator::XYZDelaunayGenerator(const InputParameters & parameters)
+  : MeshGenerator(parameters),
+    _bdy_ptr(getMesh("boundary")),
+    _desired_volume(getParam<Real>("desired_volume")),
+    _output_subdomain_id(0),
+    _smooth_tri(getParam<bool>("smooth_triangulation")),
+    _hole_ptrs(getMeshes("holes")),
+    _stitch_holes(getParam<std::vector<bool>>("stitch_holes")),
+    _algorithm(parameters.get<MooseEnum>("algorithm")),
+    _verbose_stitching(parameters.get<bool>("verbose_stitching"))
+{
+  if (!_stitch_holes.empty() && _stitch_holes.size() != _hole_ptrs.size())
+    paramError("stitch_holes", "Need one stitch_holes entry per hole, if specified.");
+}
+
+std::unique_ptr<MeshBase>
+XYZDelaunayGenerator::generate()
+{
+#ifdef LIBMESH_HAVE_NETGEN
+  // Put the boundary mesh in a local pointer
+  std::unique_ptr<UnstructuredMesh> mesh =
+      dynamic_pointer_cast<UnstructuredMesh>(std::move(_bdy_ptr));
+
+  // Get ready to triangulate its boundary
+  NetGenMeshInterface ngint(*mesh);
+
+  ngint.smooth_after_generating() = _smooth_tri;
+
+  ngint.desired_volume() = _desired_volume;
+
+  std::unique_ptr<std::vector<std::unique_ptr<UnstructuredMesh>>> ngholes =
+      std::make_unique<std::vector<std::unique_ptr<UnstructuredMesh>>>();
+
+  // The libMesh Netgen interface will modify hole meshes in-place, so
+  // we make copies to pass in.
+  for (std::unique_ptr<MeshBase> * hole_ptr : _hole_ptrs)
+  {
+    // How did we never add a ReplicatedMesh(MeshBase&) constructor in
+    // libMesh?
+    const UnstructuredMesh & hole = dynamic_cast<UnstructuredMesh &>(**hole_ptr);
+    ngholes->push_back(std::make_unique<ReplicatedMesh>(hole));
```

Comment:
will this serialize/replicate or should we be checking they are replicated already?

Reply:
The `ReplicatedMesh(UnstructuredMesh&)` constructor handles it ... as of https://github.com/libMesh/libmesh/pull/3921 anyway.